File: Diffusion/Project1_code/Project1.cpp
// use point SOR to solve the finite-volume method "Del^2 phi = 0"
// Ax=b
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

class node{
    public:
        double x,y,n,phi,phiN;  // x,y coordinates,node number, and phi value, and next iteration phi value
        double N,E,W,S; // neighbors in North, East, West, and South directions for 2d
        node(){             // constructor of 0's
            x       =   0.  ;
            y       =   0.  ;
            n       =   0.  ;
            phi     =   0.  ;
            phiN    =   0.  ;
        }


};
void show(vector< vector<node> > a){
    cout<<" x,y,n,phi,phiN for each position = "<<endl;
    for (auto nodes_it = a.begin(); nodes_it!=a.end(); ++nodes_it){
        for (auto  nodes_it_j = nodes_it->begin(); nodes_it_j!=nodes_it->end(); ++nodes_it_j){
            cout
                <<nodes_it_j->x<<","
                <<nodes_it_j->y<<","
                <<nodes_it_j->n<<","
                <<nodes_it_j->phi<<","
                <<nodes_it_j->phiN<<","
                <<"  ";
        }
        cout<<"end line"<<endl;
    }
}

int main(){
    vector< vector<node> > nodes(20,vector<node> (20));       // 20 by 20 matrix containing node data
    vector< vector<double> > A,D,L,U;   // A matrix and it's subcomponents D=diagonal,L=lower,U=upper (A=D + L + U)
    vector<double> x,b;
    vector<double> aw,ae,as,an,ap;         // coefficients on west, east, south and north neighbors of center point
    vector< vector<double> >::iterator i;
    vector<double>::iterator j;
    double temp;
    double w;       // relaxation factor for SOR method
    double dx=1./20.,dy=1./20.;     // spacing

    int MAX_R=20,MAX_C=20;    // 20 by 20 finite volume 2d problem

    // set node values
    int xi=0, yj=0, nn=0;
    for (auto nodes_it = nodes.begin(); nodes_it!=nodes.end(); ++nodes_it,++yj){
        xi=0;
        for (auto  nodes_it_j = nodes_it->begin(); nodes_it_j!=nodes_it->end(); ++nodes_it_j,++xi){
            nodes_it_j->n   = nn++;
            nodes_it_j->phi = 0.;
            nodes_it_j->x   = xi*dx+dx/2.;
            nodes_it_j->y   = yj*dy+dy/2.;
            if (xi*dx==1.) nodes_it_j->phi = yj;
            if (yj*dx==1.) nodes_it_j->phi = xi;
        }
    }
    // show nodes
    show(nodes);
    // set aw, ae, as, and an, and ap values 
    ap.push_back(2.+0.);    // top left BC
    for (int xi=0;xi<MAX_R; xi++){
        for (int yj=0;yj<MAX_C; yj++){
            aw.push_back(1.);
            ae.push_back(1.);
            as.push_back(1.);
            an.push_back(1.);
            ap.push_back(aw[xi]+ae[xi]+as[xi]+an[xi]);
        }
    }
    // set BC values on RHS and diagonal(ap)



    //initialize values of A (D,L, and U subcomponents) and b
    /*        for (int xi=0; xi<MAX_R; xi++){
              A.push_back(vector<double> ());
              for (int yj=0; yj<MAX_C; yj++){
              A[xi].push_back(double(xi*yj));
    //            cout<<A[xi][yj]<<" ";
    }
    //        cout<<endl;
    }
    //    cout<<endl<<endl;
    // split into subcomponents
    xi=0, yj=0;
    for (i=A.begin(); i!=A.end(); i++,xi++){
    D.push_back(vector<double> ());
    U.push_back(vector<double> ());
    L.push_back(vector<double> ());
    yj = 0;
    for (j=i->begin(); j!=i->end(); j++,yj++){
    temp = *j;
    if (xi == yj){              // diagonal
    D[xi].push_back(temp);
    U[xi].push_back(0.);
    L[xi].push_back(0.);
    }
    else if (xi < yj) {         // upper
    D[xi].push_back(0.);
    U[xi].push_back(*j);
    L[xi].push_back(0.);
    }
    else if (xi > yj){          // lower
    D[xi].push_back(0.);
    U[xi].push_back(0.);
    L[xi].push_back(*j);
    }
    else 
    cout<<"something is wrong"<<endl;

    //            cout<<*j<<" ";
    }
    //        cout<<endl;
    }
    // output to check
    //    cout<<"D = "<<endl;
    //    for (xi=0;xi<MAX_R;++xi){
    //        for (yj=0;yj<MAX_C;++yj){
    //            cout<<D[xi][yj]<<" ";
    //        }
    //        cout<<endl;
    //    }
    //    cout<<"U = "<<endl;
    //    for (xi=0;xi<MAX_R;++xi){
    //        for (yj=0;yj<MAX_C;++yj){
    //            cout<<U[xi][yj]<<" ";
    //        }
    //        cout<<endl;
    //    }
    //    cout<<"L = "<<endl;
    //    for (xi=0;xi<MAX_R;++xi){
    //        for (yj=0;yj<MAX_C;++yj){
    //            cout<<L[xi][yj]<<" ";
    //        }
    //        cout<<endl;
    //    }



    // solve x values using SOR method


*/
    return 0;
}

File: Diffusion/Project1_code/Project1.f90
! user defined variables to define finite volume
! x and y direction # of cells
#define max_x 20
#define max_y 20
! x and y number of cells plus 1
#define max_xp 21
#define max_yp 21
! x and y number of cells plus 2 (to account for boundary nodes)
#define max_x2p 22
#define max_y2p 22
MODULE types
    !purpose: define data type struct
    IMPLICIT NONE
    TYPE::dat
        REAL::x,y,phi,phi_new
        INTEGER::n
    END TYPE dat
CONTAINS
    SUBROUTINE set_xy (strct,dx,dy,nx,ny)
        REAL,INTENT(IN)     ::  dx,dy
        INTEGER,INTENT(IN)  ::  nx,ny   ! size of strct in x and y directions 
        TYPE(dat),DIMENSION(0:nx-1,0:nx-1),INTENT(INOUT)::strct ! data contained from 0:nx-1 where cells 0 and nx-1 are boundary nodes (cell volume approaches 0 on boundary nodes)
        INTEGER ::  i,j,n   ! for do loops and n is counter for cell number
        REAL    ::  xi,yi   ! x and y values for each cell
        n=1                     ! cell number 1
        DO i=1,ny-2             ! 1 to ny-2 for boundary nodes (we only are iterating through the middle values)
            yi = i*dy - dy/2.   ! y coordinate
            DO j=1,nx-2
                xi = j*dx - dx/2.       ! x coordinate
                strct(i,j)%n = n        ! input n node
                strct(i,j)%x = xi       ! x coordinate to strct
                strct(i,j)%y = yi       ! y coordinate to strct
                strct(i,j)%phi = 0.     ! phi value initialized guess
                strct(i,j)%phi_new = 0. ! phi value initialized guess for next iteration
                n=n+1                   ! count cell numbers up one
            END DO
        END DO
    END SUBROUTINE set_xy
END MODULE types

PROGRAM project1
    USE types !use module defined by types
    IMPLICIT NONE
    ! declare variables
    INTEGER ::  i,j,iter!,max_x=20,max_y=20
    REAL    ::  dx,dy,gamma,deltaPhi,ae,aw,an,as,ap,error
    TYPE(dat),DIMENSION(0:max_xp,0:max_yp)::phi ! 22 if you count edges (thin cell)
    ! set dx and dy and gamma and coefficients (without dividing by delta x between node centers)
    dx=1./REAL(max_x)
    dy=1./REAL(max_y)
    gamma = 1.
    ! initialize phi data and x,y for middle values
    CALL set_xy(phi,dx,dy,max_x2p,max_x2p)
    !! initialize BC's
    ! left Boundary
    phi(:,0)%x = 0.
    phi(0,0)%y = 0.
    phi(1:max_x,0)%y = RESHAPE((/ (i*dy - dy/2. ,i=1,max_x) /),(/ max_x /))
    phi(max_xp,0)%y = 1.
    phi(:,0)%phi = 0.
    phi(:,0)%phi_new = 0.
    ! top boundary
    phi(0,0)%x = 0.
    phi(0,1:max_y)%x = RESHAPE((/ (i*dx - dx/2. ,i=1,max_y) /),(/ max_y /))
    phi(0,max_yp)%x = 1.
    phi(0,:)%y = 0.
    phi(0,:)%phi = 0.
    phi(0,:)%phi_new = 0.
    ! right boundary
    phi(:,max_yp)%x = 1.
    phi(:,max_yp)%y = phi(:,0)%y
    phi(:,max_yp)%phi = phi(:,max_yp)%y
    phi(:,max_yp)%phi_new = phi(:,max_yp)%y
    ! bottom boundary
    phi(max_xp,:)%x = phi(0,:)%x
    phi(max_xp,:)%y = 1.
    phi(max_xp,:)%phi = phi(max_xp,:)%x
    phi(max_xp,:)%phi_new = phi(max_xp,:)%x
    ! point SOR method to solve for the exact values of phi using the BC (only loop through inner 20X20 values)
    open(unit=5,file="convergence.txt")
    DO iter=0,100000
        error = 0.
        DO i=max_y,1,-1
            DO j=max_x,1,-1
                ae=dy*gamma/(abs(phi(i,j)%x-phi(i,j+1)%x)) 
                aw=dy*gamma/(abs(phi(i,j)%x-phi(i,j-1)%x)) 
                an=dx*gamma/(abs(phi(i,j)%y-phi(i+1,j)%y)) 
                as=dx*gamma/(abs(phi(i,j)%y-phi(i-1,j)%y)) 
                ap=(ae+aw+an+as) 
                deltaPhi = ae*phi(i,j+1)%phi_new + &    ! use new iteration values
                    aw*phi(i,j-1)%phi_new + &           ! use new iteration values
                    an*phi(i+1,j)%phi_new + &           ! use new iteration values
                    as*phi(i-1,j)%phi_new - &           ! use new iteration values
                    ap*phi(i,j)%phi                     ! use old iteration values
                phi(i,j)%phi_new = phi(i,j)%phi + 1.8/ap * deltaPhi     ! each iterations approximate new phi value
                error = error + deltaPhi**2             ! root sum square the error for each iteration
            END DO
        END DO
        error = SQRT(error)                             ! sqrt to have RSS of error
        write(5,*) iter,error
        phi%phi = phi%phi_new                           ! set old phi to the new iteration guess
        IF (error .lt. 1.E-14) THEN
            WRITE(*,*) 'Did in ',iter,' iterations'
            WRITE(*,*) 'With RSS error = ',error
            EXIT                                        ! exit loop if error is small enough
        END IF
    END DO
    !output
    ! user will need to specify size of 
    open(unit=9,file="x.txt");open(unit=10,file="y.txt");open(unit=11,file="phi.txt");open(unit=12,file="error.txt")
    100 FORMAT (max_x2p F10.6)
    200 FORMAT (max_x2p ES16.6)
    WRITE(9,100) ( phi(i,:)%x ,i=0,21 )
    WRITE(10,100) ( phi(i,:)%y ,i=0,21 )
    WRITE(11,100) ( phi(i,:)%phi ,i=0,21 )
    WRITE(12,200) ( abs(phi(i,:)%phi - (phi(i,:)%x*phi(i,:)%y)),i=0,21 )
    close(9);close(10);close(11);close(12);close(5);
END PROGRAM project1
